File: test_regress/t/t_dist_portability.pl
#!/usr/bin/env perl
if (!$::Driver) { use FindBin; exec("$FindBin::Bin/bootstrap.pl", @ARGV, $0); die; }
# DESCRIPTION: Verilator: Verilog Test driver/expect definition
#
# Copyright 2003 by Wilson Snyder. This program is free software; you
# can redistribute it and/or modify it under the terms of either the GNU
# Lesser General Public License Version 3 or the Perl Artistic License
# Version 2.0.
# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0

use IO::File;

scenarios(dist => 1);

my $root = "..";
my $Debug;

if (!-r "$root/.git") {
    skip("Not in a git repository");
} else {
    printfll();
    cstr();
    vsnprintf();
    final();
}

ok(1);

sub printfll {
    my $files = "src/*.c* src/*.h include/*.c* include/*.h examples/*/*.c* test_regress/t/*.c* test_regress/t/*.h";
    my $cmd = "cd $root && fgrep -n ll $files | sort";
    print "C $cmd\n";
    my $grep = `$cmd`;
    my %names;
    foreach my $line (split /\n/, $grep) {
        next if $line !~ /%[a-z0-9]*ll/;
        next if $line !~ /\blong\S+long\b/;  # Assume a cast
        print "$line\n";
        if ($line =~ /^([^:]+)/) {
            $names{$1} = 1;
        } else {
            $names{UNKNOWN} = 1;
        }
    }
    if (keys %names) {
        error("Files with %ll instead of PRIx64: ", join(' ', sort keys %names));
    }
}

sub cstr {
    my $files = "src/*.c* src/*.h include/*.c* include/*.h examples/*/*.c* test_regress/t/*.c* test_regress/t/*.h";
    my $cmd = "cd $root && grep -n -P 'c_str|begin|end' $files | sort";
    print "C $cmd\n";
    my $grep = `$cmd`;
    my %names;
    foreach my $line (split /\n/, $grep) {
        if ($line =~ /^([^:]+)[^"]*\(\)[a-z0-9_().->]*[.->]+(c_str|r?begin|r?end)\(\)/) {
            next if $line =~ /lintok-begin-on-ref/;
            print "$line\n";
            $names{$1} = 1;
        }
    }
    if (keys %names) {
        error("Files with potential c_str() lifetime issue: ", join(' ', sort keys %names));
    }
}

sub vsnprintf {
    # Note do not do test_regress, as VPI files need to compile without verilatedos.h
    my $files = "src/*.c* src/*.h include/*.c* include/*.h examples/*/*.c*";
    my $cmd = "cd $root && grep -n -P '(snprintf|vsnprintf)' $files | sort";
    print "C $cmd\n";
    my $grep = `$cmd`;
    my %names;
    foreach my $line (split /\n/, $grep) {
        if ($line =~ /\b(snprintf|vsnprintf)\b/) {
            next if $line =~ /# *define\s*VL_V?SNPRINTF/;
            print "$line\n";
            $names{$1} = 1;
        }
    }
    if (keys %names) {
        error("Files with vsnprintf, use VL_VSNPRINTF: ", join(' ', sort keys %names));
    }
}

sub final {
    # Note do not do test_regress, as VPI files need to compile without verilatedos.h
    my $files = "src/*.c* src/*.h include/*.c* include/*.h";
    my $cmd = "cd $root && grep -n -P '(class)' $files | sort";
    print "C $cmd\n";
    my $grep = `$cmd`;
    my %names;
    foreach my $line (split /\n/, $grep) {
        if ($line =~ /:\s*class /) {
            next if $line =~ /final|VL_NOT_FINAL/;
            next if $line =~ /{}/;  # e.g. 'class Foo {};'
            next if $line =~ /;/;  # e.g. 'class Foo;'
            print "$line\n";
            $names{$1} = 1;
        }
    }
    if (keys %names) {
        error("Files with classes without final/VL_NOT_FINAL: ", join(' ', sort keys %names));
    }
}

1;

File: test_regress/t/t_dist_portability.py
#!/usr/bin/env python3
# DESCRIPTION: Verilator: Verilog Test driver/expect definition
#
# Copyright 2024 by Wilson Snyder. This program is free software; you
# can redistribute it and/or modify it under the terms of either the GNU
# Lesser General Public License Version 3 or the Perl Artistic License
# Version 2.0.
# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0

import vltest_bootstrap

test.scenarios('dist')

root = ".."


def line_filename(line):
    m = re.search(r'^([^:]+)', line)
    return m.group(1) if m else 'UNKNOWN'


def printfll():
    files = "src/*.c* src/*.h include/*.c* include/*.h examples/*/*.c* test_regress/t/*.c* test_regress/t/*.h"
    cmd = "cd " + root + " && fgrep -n ll " + files + " | sort"
    grep = test.run_capture(cmd, check=False)
    names = {}
    for line in grep.splitlines():
        if not re.search(r'%[a-z0-9]*ll', line):
            continue
        if re.search(r'lintok-format-ll', line):
            continue
        print(line)
        names[line_filename(line)] = True
    if len(names):
        test.error("Files with %ll instead of PRIx64: " + ' '.join(sorted(names.keys())))


def cstr():
    files = "src/*.c* src/*.h include/*.c* include/*.h examples/*/*.c* test_regress/t/*.c* test_regress/t/*.h"
    cmd = "cd " + root + " && grep -n -P 'c_str|begin|end' " + files + " | sort"
    grep = test.run_capture(cmd, check=False)
    names = {}
    for line in grep.splitlines():
        if re.search(r'^([^:]+)[^"]*\(\)[a-z0-9_().->]*[.->]+(c_str|r?begin|r?end)\(\)', line):
            if re.search(r'lintok-begin-on-ref', line):
                continue
            print(line)
            names[line_filename(line)] = True
    if len(names):
        test.error("Files with potential c_str() lifetime issue: " +
                   ' '.join(sorted(names.keys())))


def vsnprintf():
    # Note do not do test_regress, as VPI files need to compile without verilatedos.h
    files = "src/*.c* src/*.h include/*.c* include/*.h examples/*/*.c*"
    cmd = "cd " + root + " && grep -n -P '(snprintf|vsnprintf)' " + files + " | sort"
    grep = test.run_capture(cmd, check=False)
    names = {}
    for line in grep.splitlines():
        if re.search(r'\b(snprintf|vsnprintf)\b', line):
            if re.search(r'# *define\s*VL_V?SNPRINTF', line):
                continue
            print(line)
            names[line_filename(line)] = True
    if len(names):
        test.error("Files with vsnprintf, use VL_VSNPRINTF: " + ' '.join(sorted(names.keys())))


def final():
    # Note do not do test_regress, as VPI files need to compile without verilatedos.h
    files = "src/*.c* src/*.h include/*.c* include/*.h"
    cmd = "cd " + root + " && grep -n -P '(class|struct)' " + files + " | sort"
    grep = test.run_capture(cmd, check=False)
    names = {}
    for line in grep.splitlines():
        if re.search(r':\s*(class|struct) ', line):
            if (re.search(r'final|VL_NOT_FINAL', line)  #
                    or re.search(r'{}', line)  # e.g. 'class Foo {};'
                    or re.search(r';', line)  # e.g. 'class Foo;'
                    or re.search(r'(class|struct)\s+{', line)  # e.g. anon 'class {'
                    or re.search(r'struct std::', line)  #
                    or not re.search(r'{', line)):
                continue
            print(line)
            names[line_filename(line)] = True
    if len(names):
        test.error("Files with classes without final/VL_NOT_FINAL: " +
                   ' '.join(sorted(names.keys())))


if not os.path.exists(root + "/.git"):
    test.skip("Not in a git repository")

printfll()
cstr()
vsnprintf()
final()

test.passes()
